<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.32">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>GWAS Resources â€“ Data Science for Health and Biomedical Sciences</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../..//favicon.png" rel="icon" type="image/png">
<script src="../../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-37eea08aefeeee20ff55810ff984fec1.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap-c8db1e2376186e53c21d3da13e85f734.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Data Science for Health and Biomedical Sciences</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-course-information" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Course Information</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-course-information">    
        <li>
    <a class="dropdown-item" href="../../course-information/welcome.html">
 <span class="dropdown-text">Welcome and Learning Outcomes</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../course-information/index.html">
 <span class="dropdown-text">Course Information</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../course-information/timetable.html">
 <span class="dropdown-text">Timetable</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../course-information/library-resources.html">
 <span class="dropdown-text">Library Resources</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../course-information/course-contacts.html">
 <span class="dropdown-text">Course Contacts</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../course-information/gen-ai-guidance.html">
 <span class="dropdown-text">Gen AI Guidance</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../course-information/help-and-support.html">
 <span class="dropdown-text">Help and Support</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="../../content/index.html"> 
<span class="menu-text">Content</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../assessment/index.html"> 
<span class="menu-text">Assessment</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#software" id="toc-software" class="nav-link active" data-scroll-target="#software">Software</a></li>
  <li><a href="#summary-statistics" id="toc-summary-statistics" class="nav-link" data-scroll-target="#summary-statistics">Summary statistics</a>
  <ul class="collapse">
  <li><a href="#sources" id="toc-sources" class="nav-link" data-scroll-target="#sources"><strong>Sources</strong></a></li>
  <li><a href="#file-structure" id="toc-file-structure" class="nav-link" data-scroll-target="#file-structure"><strong>File structure</strong></a></li>
  <li><a href="#file-structure-1" id="toc-file-structure-1" class="nav-link" data-scroll-target="#file-structure-1"><strong>File structure</strong></a></li>
  </ul></li>
  <li><a href="#genomic-databases" id="toc-genomic-databases" class="nav-link" data-scroll-target="#genomic-databases">Genomic databases</a></li>
  <li><a href="#suggested-read-gwas-paper" id="toc-suggested-read-gwas-paper" class="nav-link" data-scroll-target="#suggested-read-gwas-paper">Suggested Read: GWAS Paper</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">GWAS Resources</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="software" class="level2">
<h2 class="anchored" data-anchor-id="software">Software</h2>
<p>A number of software packages have been developed for conducting genome-wide association studies. A few of the more common packages are:</p>
<ul>
<li><p><a href="https://www.cog-genomics.org/plink/1.9/">PLINK</a>: General-purpose command line program available on Linux, Windows, and MacOS. Contains routines for opening, filtering, and converting genomic data in a number of formats. Runs linear and logistic association tests for unrelated participants. Also has a <a href="https://www.cog-genomics.org/plink/2.0/">version 2</a> under development. This is the program that is taught in most GWAS courses.</p></li>
<li><p><a href="https://yanglab.westlake.edu.cn/software/gcta">GCTA</a>: Software package primarily for estimating heritabilty using genomic data, it can also perform GWAS while correcting for relatedness among participants. Contians a number of other features for downstream analysis, such as conditional analysis and gene-based tests. Uses a linear approximation for case-control GWAS.</p></li>
<li><p><a href="https://dougspeed.com/">LDAK</a>: Similar to GCTA but can fit models for different genetic architectures.&nbsp;</p></li>
<li><p><a href="https://alkesgroup.broadinstitute.org/BOLT-LMM/BOLT-LMM_manual.html">BOLT-LMM</a>: Similar to GCTA but uses a fast approximation that speeds up computation considerably. Appropriate when analysing very large samples.&nbsp;Uses a linear approximation for case-control GWAS.</p></li>
<li><p><a href="https://rgcgithub.github.io/regenie/">regenie</a>: Fast software when GWASing multiple traits, possibly with related participants. Applicable to case-control GWAS.</p></li>
</ul>
<p>There are several R packages that are useful for GWAS:</p>
<ul>
<li><p><a href="https://privefl.github.io/bigsnpr">bigsnpr</a>: Managing and performing quality control on SNP data</p></li>
<li><p><a href="https://bioconductor.org/">Bioconductor</a>: repository of bioinformatics packages</p></li>
<li><p><a href="https://github.com/MRCIEU/ieugwasr">ieugwasr</a>: query the OpenGWAS database</p></li>
</ul>
<p>Web applications:</p>
<ul>
<li><p><a href="https://fuma.ctglab.nl/">FUMA</a>: functional annotation and mapping of GWAS summary statistics</p></li>
<li><p><a href="http://locuszoom.org/">LocusZoom</a>: interactive region plots from GWAS summary statistics</p></li>
</ul>
<p>Additional tools for different stages of a GWAS analysis are listed in <a href="https://www.nature.com/articles/s43586-021-00056-9/tables/1">Table 1 of the Uffelmann primer</a>.</p>
<p>There is also a big list of software used for genetic analysis maintained on GitHub as the Rockefeller List: <a href="https://gaow.github.io/genetic-analysis-software/0/" class="uri">https://gaow.github.io/genetic-analysis-software/0/</a></p>
</section>
<section id="summary-statistics" class="level2">
<h2 class="anchored" data-anchor-id="summary-statistics">Summary statistics</h2>
<section id="sources" class="level3">
<h3 class="anchored" data-anchor-id="sources"><strong>Sources</strong></h3>
<p>Summary statistics are the output of SNP association effect sizes and p-values from a GWAS. Summary statistics are available in full or in part from a number of websites:</p>
<ul>
<li><p><a href="https://www.ebi.ac.uk/gwas/">GWAS Catalog</a>: Comprehensive catalog of phenotype-SNP associations. Usually only contains independent significant SNPs from each study.</p></li>
<li><p><a href="https://gwas.mrcieu.ac.uk/">OpenGWAS</a>: Database of full summary statistics for a large number of traits, available for querying and downloading in a standardised format.</p></li>
<li><p><a href="https://pgc.unc.edu/for-researchers/download-results/">Psychiatric Genomics Consortium (PGC)</a>: Full summary statistics on psychiatric disorders.&nbsp;</p></li>
<li><p><a href="https://www.thessgac.org/">Social Sciences Genetic Association Consortium (SSGAC)</a>: Full summary statistics on social and behavioural traits like education and personality.</p></li>
<li><p><a href="https://www.nealelab.is/uk-biobank">Nealelab</a>: Transancestry GWAS from UK Biobank.</p></li>
<li><p><a href="https://portals.broadinstitute.org/collaboration/giant/index.php/GIANT_consortium_data_files">GIANT Consortium</a>: GWAS of anthropomorphic phenotypes (weight and height).</p></li>
<li><p><a href="http://lipidgenetics.org/">Global Lipids Genetics Consortium</a>: GWAS of lipids.</p></li>
</ul>
</section>
<section id="file-structure" class="level3">
<h3 class="anchored" data-anchor-id="file-structure"><strong>File structure</strong></h3>
<p>Summary statistics are typically provided as a plain-text, tabular data file (columns are usually space or tab delimited). There is no single standard format (for example, see <a href="https://docs.google.com/document/d/1TWIhr8-qpCXB13WCXcU1_HDio8lC_MeWoAg2jlggrtU/edit">daner</a>, <a href="https://github.com/MRCIEU/gwas-vcf-specification">GWAS-VCF,</a> and <a href="https://ebispot.github.io/gwas-blog/new-standard-for-gwas-summary-statistics">GWAS-SSF</a>), but most files will look something like the following:</p>
</section>
<section id="file-structure-1" class="level3">
<h3 class="anchored" data-anchor-id="file-structure-1"><strong>File structure</strong></h3>
<p>Summary statistics are typically provided as a plain-text, tabular data file (columns are usually space or tab delimited). There is no single standard format (for example, see <a href="https://docs.google.com/document/d/1TWIhr8-qpCXB13WCXcU1_HDio8lC_MeWoAg2jlggrtU/edit">daner</a>, <a href="https://github.com/MRCIEU/gwas-vcf-specification">GWAS-VCF,</a> and <a href="https://ebispot.github.io/gwas-blog/new-standard-for-gwas-summary-statistics">GWAS-SSF</a>), but most files will look something like the following:</p>
<table class="caption-top table">
<colgroup>
<col style="width: 7%">
<col style="width: 18%">
<col style="width: 10%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 12%">
<col style="width: 12%">
<col style="width: 12%">
<col style="width: 11%">
</colgroup>
<tbody>
<tr class="odd">
<td>CHR</td>
<td>SNP</td>
<td>POS</td>
<td>REF</td>
<td>ALT</td>
<td>AF</td>
<td>BETA</td>
<td>SE</td>
<td>P</td>
</tr>
<tr class="even">
<td>1</td>
<td>rs189107123</td>
<td>10611</td>
<td>C</td>
<td>G</td>
<td>0.985&nbsp;</td>
<td>-0.128</td>
<td>0.0761</td>
<td>0.0926</td>
</tr>
<tr class="odd">
<td>1</td>
<td>rs180734498</td>
<td>13302</td>
<td>T&nbsp;&nbsp;</td>
<td>C&nbsp;&nbsp;</td>
<td>0.118&nbsp;</td>
<td>0.0389&nbsp;</td>
<td>0.0278</td>
<td>0.162</td>
</tr>
<tr class="even">
<td>1</td>
<td>rs144762171</td>
<td>13327</td>
<td>C&nbsp;&nbsp;</td>
<td>G&nbsp;&nbsp;</td>
<td>0.0339&nbsp;</td>
<td>0.0417</td>
<td>&nbsp;0.049&nbsp;</td>
<td>0.394</td>
</tr>
<tr class="odd">
<td>â€¦</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
</tr>
</tbody>
</table>
<ul>
<li><p>CHR: the chromosome number of symbol. Usually specified as a number (1, 2, 3, â€¦) for GCRh37 or prepended with â€˜chrâ€™ for GCRh38 (chr1, chr2, chr3, â€¦)</p></li>
<li><p>SNP: <a href="https://www.ncbi.nlm.nih.gov/snp/docs/RefSNP_about/">Reference SNP number</a> (RSID) for the genetic marker, or some other unique name. Sometimes listed just using the chromsome and base pair position (CPID) identifier with out without the alleles (e.g., SNP rs189107123 might be called 1:10611 or 1:10611_C_G in some datasets)</p></li>
<li><p>POS: basepair position in genomic coordinates of the reference genome build. This column is sometimes also labelled as BP.</p></li>
<li><p>REF: reference allele. Usually, but not always, the effect allele for the association. This column is also sometimes labelled A1, allele1, or alleleB</p></li>
<li><p>ALT: alternative allele. Usually, but not always, the non-effect allele. This column might also be called A2 or allele0.</p></li>
<li><p>AF: frequency of the reference allele. May be multiple columns if allele frequencies in cases and controls are reported seperately.</p></li>
<li><p>BETA: effect size of the association (substitution effect of each additional copy of the effect allele). For case/control and binary phenotypes, usually the odds ratio (OR) is reported instead.</p></li>
<li><p>SE: standard error of the effect size. When the effect size is an odds ratio, this column usually represents the standard error of log(OR)</p></li>
<li><p>P: p-value of the association</p></li>
</ul>
<p>Depending on the software and study design, additional columns might be present or missing.</p>
<p>The three most important points to check when processing GWAS summary statistics are:</p>
<ul>
<li><p>what genome build the summary statistics used? RSIDs and basepair coordinates differ between genome builds, so this can influence merging and look up with other tools.</p></li>
<li><p>which allele is the effect allele? There is inconsistency in whether the REF or ALT allele is the effect allele, and even how these columns are labelled. (<a href="https://academic.oup.com/ije/article/49/5/1734/5900777">â€œLetâ€™s call it the effect allele</a>â€œ)</p></li>
<li><p>which strand do the alleles refer to? While the positive strand is usually assumed, some datasets might mix in alleles coded from the reverse complementary strand of DNA (e.g., the alleles for rs180734498 might be reported as REF=A, ALT=G). This causes problems when the two alleles are complementary (C and G, A and T).</p></li>
</ul>
</section>
</section>
<section id="genomic-databases" class="level2">
<h2 class="anchored" data-anchor-id="genomic-databases">Genomic databases</h2>
<p>The <a href="https://www.ncbi.nlm.nih.gov/">National Center for Biotechnology Information (NCBI)</a>, run by NIH, has a number of specialised, cross-referenced databases for querying genomic information:</p>
<ul>
<li><p><a href="https://www.ncbi.nlm.nih.gov/snp/">dbSNP</a>: database of SNP variants with information on population frequency and genomic context.</p></li>
<li><p><a href="https://www.ncbi.nlm.nih.gov/gene/">Gene</a>: cross species information on genes.</p></li>
<li><p><a href="https://www.ncbi.nlm.nih.gov/clinvar/">ClinVar</a>: Genetic variation with clinical relevance to human health</p></li>
</ul>
<p>Other useful resources include</p>
<ul>
<li><p><a href="https://omim.org/">OMIM</a>: Online Catalog of Human Genes and Genetic Disorders</p></li>
<li><p><a href="http://www.gnomad-sg.org/">genomeAD</a>: Aggregation database of exome and genome sequencing datasets.&nbsp;</p></li>
</ul>
</section>
<section id="suggested-read-gwas-paper" class="level2">
<h2 class="anchored" data-anchor-id="suggested-read-gwas-paper">Suggested Read: GWAS Paper</h2>
<p><a href="files/Letâ€™s_call_it_the effect allele_a_suggestion_for_GWAS_naming_conventions.pdf">Link</a></p>
<hr>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "î§‹";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/ddi-talent\.github\.io\/website-ug-data-science\/");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
            // target, if specified
            link.setAttribute("target", "_blank");
            if (link.getAttribute("rel") === null) {
              link.setAttribute("rel", "noopener");
            }
            // default icon
            link.classList.add("external");
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>