---
title: "Genome-wide association studies"
format: html
---

## Performing an association test

![](images/week10.3.SNP_association_test.png)

An association test is performed by comparing a genotype at a particular SNP marker to a phenotype (a trait or disease). This test is usually done within a regression framework, using linear regression for continuous/quantitative phenotypes (like height) or a logistic regression for a binary or case/control phenotype (like disease status). 

At each SNP there are typically two alleles, for example C and T*.* This allows for four possible genotypes: CC homozygote, CT or TC heterozyogotes, and TT homozygotes. The basic association tests assumes that substituting one allele for the other has an additive effect on the phenotype. This means that the model assumes that the average difference in phenotype between the CC genotype and the CT genotype is the same as between the CT and TT genotype.

To run the regression, the SNP genotype is encoded to a number as a count of one of the alleles (which allele is chosen is arbitrary, but traditionally it is the minor allele). The chosen allele is called the "effect allele". In this example, if the T allele is chosen as the effect allele, then the numeric encoding is the number of T alleles in the genotype: CC = 0, CT and TC = 1, and TT = 2. This numeric representation can then be entered into a regression formula:

![](images/week10.3.regression_formula.png){fig-align="center"}

 where *Y~i~* is the phenotype of participant *i*, ***X~i~*** is a vector of covariates for participant *i*, *g~ij~* is the numerical genotype for participant *i* at SNP *j*, and *e~ij~* is the error term. **B** is a vector of regression coefficients for the covariates and *β~j~* is the regression coefficient for SNP *j*. From the fitted regression model, the beta coefficient *β~j~* for the SNP represents the average substitution effect of having one additional copy of the effect allele. 

Binary traits like disease status are analysed as case-control studies using logistic regression. In a case-control analysis, the beta coefficient represents the change in log-odds of having the disease for every additional copy of the effect allele. Exponentiating the beta coefficient,  [![e to the power of straight beta blank subscript j end exponent](data:image/svg+xml;charset=utf8,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20xmlns%3Awrs%3D%22http%3A%2F%2Fwww.wiris.com%2Fxml%2Fmathml-extension%22%20height%3D%2226%22%20width%3D%2220%22%20wrs%3Abaseline%3D%2221%22%3E%3C!--MathML%3A%20%3Cmath%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F1998%2FMath%2FMathML%22%3E%3Cmsup%3E%3Cmi%3Ee%3C%2Fmi%3E%3Cmsub%3E%3Cmi%3E%26%23x3B2%3B%3C%2Fmi%3E%3Cmi%3Ej%3C%2Fmi%3E%3C%2Fmsub%3E%3C%2Fmsup%3E%3C%2Fmath%3E--%3E%3Cdefs%3E%3Cstyle%20type%3D%22text%2Fcss%22%2F%3E%3C%2Fdefs%3E%3Ctext%20font-family%3D%22Times%20New%20Roman%22%20font-size%3D%2216%22%20font-style%3D%22italic%22%20text-anchor%3D%22middle%22%20x%3D%223.5%22%20y%3D%2221%22%3Ee%3C%2Ftext%3E%3Ctext%20font-family%3D%22Times%20New%20Roman%22%20font-size%3D%2212%22%20font-style%3D%22italic%22%20text-anchor%3D%22middle%22%20x%3D%2211.5%22%20y%3D%2211%22%3E%26%23x3B2%3B%3C%2Ftext%3E%3Ctext%20font-family%3D%22Times%20New%20Roman%22%20font-size%3D%2210%22%20font-style%3D%22italic%22%20text-anchor%3D%22middle%22%20x%3D%2216.5%22%20y%3D%2216%22%3Ej%3C%2Ftext%3E%3C%2Fsvg%3E){alt="e to the power of straight beta blank subscript j end exponent"}]{.underline} , converts the effect size to an odds ratio.

## Population stratification

One possible confounder when performing association tests is that many allele frequencies differ between populations and subpopulations, while disease prevalences also differ. This can create false associations between SNPs and diseases. 

These systematic differences in allele frequencies between subpopulations is called "population stratification". The most obvious cause of stratification is ancestry but it can also be caused by confounding between environmental factors and geography. A principal components analysis of genetic variation across the whole genome can be used to visualise this variation. 

A statistical summary of the first two major dimensions of variation (PC = principal component). Each participant is plotted at their genetic coordinates using text label for their country of origin. The plot resembles the geographic relationship between each country.

![](images/week10.3.Population_stratification.png){fig-align="center"}

Novembre et al "[Genes mirror geography within Europe](https://www.nature.com/articles/nature07331)" 2008

The association between polygenic burden for traits and diseases and UK geography can be explored on the [Genes & Geography in Great Britain](https://holtzyan.shinyapps.io/UKB_geo/) web app, based on the paper Abdellaoui *et al.* "[Genetic correlates of social stratification in Great Britain](https://doi.org/10.1038/s41562-019-0757-5)" 2019.

There are other sources of false positives, such as technical artifacts and batch effects when cases and controls are genotyped separately, but these are usually easier to detect and remove during standard quality control (QC) processing. 

## Multiple testing and significanceac

Genetic data can contain hundreds of thousands or millions of variants. What p-value should be used to determine statistical significance?

Standard methods to correct for multiple testing, such as Bonferroni correction based on the number of tests, are too conservative. This is because SNP markers are not independent. Instead, SNPs that are physically close to each other on a chromosome tend to be correlated (termed "linkage disequilibrium"). This correlation is because recombination breaks up the linkage between a mutation and the rest of the chromosome as it is passed down through generations. 

![](images/week10.3.snp_recombination.png){fig-align="center"}

Hettiarachchi, G., Komar, A.A. (2022). GWAS to Identify SNPs Associated with Common Diseases and Individual Risk: Genome Wide Association Studies (GWAS) to Identify SNPs Associated with Common Diseases and Individual Risk. <https://doi.org/10.1007/978-3-031-05616-1_4>

The p-value threshold can be derived empirically from the number of independent genomic regions, which has been determined to be about 1 million. Thus a multiple testing correction of p \<= 0.05/1,000,000 = 5 × 10^-8^ is used as the standard significance threshold in GWAS.

The lack of independence among SNPs also means that, for any genomic region found to be associated with the trait, it is necessary to select one or more SNPs that represent or cause the signal in that region ("clumping").

## Summarising the results

The results from a genome-wide association study can be visualised using a graph that is called a Manhattan plot. The plot features results for each SNP plotted by chromosome and basepair position on the x-axis versus transformed p-values on the y-axis. The p-values are transformed as the negative log 10 \[-log10(P)\], which turns very small p-values into large, positive numbers.

![](images/week10.3.1.Summarising_the_results.png){fig-align="center"}

[PGC SCZ 2014](https://doi.org/10.1038/nature13595).

To summarise the results, SNPs that nearby and in linkage disequilibrium with each other are clumped together, and the SNP within the clumped region with the smallest p-value is selected. There are other more advanced techniques for determining the causal SNP or set of SNPs in a region, such as [conditional-and-joint analysis](https://www.nature.com/articles/ng.2213) and [finemapping](https://stephenslab.github.io/susieR/). 

![](images/week10.3.2.Summarising_the_results.png){fig-align="center"}

Once the SNPs are selected, the GWAS signals can be integrated with other bioinformatics data to map the SNPs to genes and explore the functional implications of the associations.

![](images/week10.3.3.Summarising_the_results.png){fig-align="center"}

## Reading list

1.  [Link 1](files/Balding_Association_studies_tutorial_Nature_Reviews_Genetics_2008.pdf)

2.  [Link 2](files/Gene–environment_correlations_across_geographic_regions_affect_genome-wide_association_studies.pdf)

3.  [Link 3](files/Primer_Genome-wide_association_studies.pdf)
